common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ErrorHandler.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.network.shuffle;
+
+import java.net.ConnectException;
+
+/**
+ * Plugs into {@link RetryingBlockFetcher} to further control when an exception should be retried
+ * and logged.
+ * Note: {@link RetryingBlockFetcher} will delegate the exception to this handler only when
+ * - remaining retries < max retries
+ * - exception is an IOException
+ */
+
+public interface ErrorHandler {
+  boolean shouldRetryError(Throwable t);
+
+  default boolean shouldLogError(Throwable t) {
+    return true;
+  }
+
+  /**
+   * A no-op error handler instance.
+   */
+  ErrorHandler NOOP_ERROR_HANDLER = t -> true;
+
+  /**
+   * The error handler for pushing shuffle blocks to remote shuffle services.
+   */
+  class BlockPushErrorHandler implements ErrorHandler {
+
+    @Override
+    public boolean shouldRetryError(Throwable t) {
+      // If it is a connection time out or a connection closed exception, no need to retry.
+      if (t.getCause() != null && t.getCause() instanceof ConnectException) {
+        return false;
+      }
+      // If the block is too late, there is no need to retry it
+      return (t.getMessage() == null ||
```

Comment:
I would not recommend relying on matching error messages, instead why not create a new exception type that extends SparkException?

Reply:
The `Too Late` or `Couldn't find opportunity` messages are part of exceptions that are created by the Shuffle Service. `SparkException` is part of core module and is not available in common/network-shuffle. 
The implementation of `MergedShuffleFileManager` which we will add with [SPARK-32916](https://issues.apache.org/jira/browse/SPARK-32916) will help better understand when the server creates these exceptions. 
The failure response of uploading the stream, block push in this case, is `RPCFailure`. With `RPCFailure` the server can send back error encoded as a String. 
An improvement that we can make here is that add a BlockPushExceptionType which is part of this errorString.
Does that sound okay?

Reply:
The ErrorHandler performs the check on the client side.
While on the server side we can generate specific exception types, these exceptions are transferred to the client side as part of `RPCFailure`, which already extracts the exception stack as a String.
If we want to generate specific exception type on the client side, we would still need to parse the exception stack string.

Reply:
I agree, you still need to match the error message string on the client side. I would define new Exception types and match the name of the Exceptions, but I think this would be only a minor improvement.

Reply:
In that case, I'd prefer to keep the way it is now and resolve this comment.
Want to push on sending out the next PR which would have much more details on the server side handling logics.
We can revisit this part during the next PR, if needed.